tt_metal/impl/dispatch/command_queue.cpp

```diff
@@ -348,24 +352,580 @@ void EnqueueWriteBufferCommand::process() {
+        // Runtime Args Cmd
+        for (const auto& [dst, transfer_info] : program.program_transfer_info.runtime_args) {
+            std::uint32_t num_packed_cmds = transfer_info.size();
+            flush.relay_inline.stride =
+                sizeof(CQPrefetchCmd) + sizeof(CQDispatchCmd) +
+                align(num_packed_cmds * sizeof(CQDispatchWritePackedUnicastSubCmd), sizeof(CQDispatchCmd));
+            for (int i = 0; i < num_packed_cmds; i++) {
+                TT_ASSERT(transfer_info[i].dst_noc_info.size() == 1, "Not supporting CoreRangeSet for runtime args");
+                flush.relay_inline.stride += transfer_info[i].data.size() * sizeof(uint32_t);
+                flush.relay_inline.stride = align(flush.relay_inline.stride, L1_ALIGNMENT);
+            }
+            // In general, length does not include alignment
+            flush.relay_inline.length = flush.relay_inline.stride - sizeof(CQPrefetchCmd);
+            flush.relay_inline.stride = align(flush.relay_inline.stride, HUGEPAGE_ALIGNMENT);
+
+            flush_ptr = (uint32_t*)&flush;
+            for (int i = 0; i < sizeof(CQPrefetchCmd) / sizeof(uint32_t); i++) {
+                this->commands.push_back(*flush_ptr++);
+            }
+            CQDispatchCmd write_packed_cmd;
+            write_packed_cmd.base.cmd_id = CQ_DISPATCH_CMD_WRITE_PACKED;
+            write_packed_cmd.write_packed.is_multicast = false;
+            write_packed_cmd.write_packed.count = num_packed_cmds;
+            write_packed_cmd.write_packed.addr = 0;
+            write_packed_cmd.write_packed.size = 0;
+
+            uint32_t* write_packed_cmd_ptr = (uint32_t*)&write_packed_cmd;
+            for (int i = 0; i < sizeof(CQDispatchCmd) / sizeof(uint32_t); i++) {
+                this->commands.push_back(*write_packed_cmd_ptr++);
+            }
+            for (int i = 0; i < num_packed_cmds * sizeof(CQDispatchWritePackedUnicastSubCmd) / sizeof(uint32_t); i++) {
+                this->commands.push_back(0);
+            }
+            if ((padding = (sizeof(CQPrefetchCmd) + sizeof(CQDispatchCmd) +
+                            num_packed_cmds * sizeof(CQDispatchWritePackedUnicastSubCmd)) %
+                           L1_ALIGNMENT) != 0) {
+                for (int j = 0; j < (L1_ALIGNMENT - padding) / sizeof(uint32_t); j++) {
+                    this->commands.push_back(0);
+                }
+            }
+            // Runtime Args Data
+            for (int i = 0; i < num_packed_cmds; i++) {
+                for (int j = 0; j < transfer_info[i].data.size(); j++) {
+                    this->commands.push_back(0);
+                }
+                if ((padding = (transfer_info[i].data.size() * sizeof(uint32_t)) % L1_ALIGNMENT) != 0) {
+                    for (int j = 0; j < (L1_ALIGNMENT - padding) / sizeof(uint32_t); j++) {
+                        this->commands.push_back(0);
+                    }
+                }
+            }
```

Comment:
seems all these loops w/ push_backs above could calculate a size and do 1 reserve

Reply:
agreed, we're still not sure what the data structures to support generating commands should look like. Almeet is suggesting maybe having some shared Prefetch + Dispatch cmds for common patterns and grabbing that to populate the commands here.